### PR TITLE
[feature] ENS Support in CLI

### PR DESCRIPTION
**Pull request by awmuncy**
Can use ENS text records to reference table names.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: test/schema.test.ts [hunk 1]
```diff
@@ -23,7 +25,9 @@ describe("commands/schema", function () {
 
   test("throws with invalid chain", async function () {
     const consoleError = spy(console, "error");
-    await yargs(["schema", "valid_9999_0"]).command(mod).parse();
+    await yargs(["schema", "valid_9999_0", "--chain", "local-tabd"])
+      .command(mod)
+      .parse();
```
Comment (asutula) [inline test/schema.test.ts:30]: So this is now testing the --chain flag, not the chain id in the table name. Maybe that is fine now since the sdk should deal with testing the chain id in the table name.

File: test/info.test.ts [hunk 2]
```diff
@@ -17,16 +17,20 @@ describe("commands/info", function () {
...
     );
   });
 
-  test("throws with invalid chain", async function () {
+  test.skip("throws with invalid chain", async function () {
     const consoleError = spy(console, "error");
-    await yargs(["info", "valid_9999_0"]).command(mod).parse();
+    await yargs(["info", "valid_9999_0", "--chain", "local-tableland"])
+      .command(mod)
+      .parse();
```
Comment (asutula) [inline test/info.test.ts:33]: Looks like this is just being skipped. Not sure if it is actually fixed.

File: package.json [hunk 3]
```diff
@@ -61,8 +67,9 @@
     "typescript": "^4.8.4"
   },
   "dependencies": {
-    "@tableland/sdk": "^4.0.0-pre.4",
-    "@tableland/sqlparser": "^1.0.0",
+    "@ensdomains/ensjs": "3.0.0-alpha.39",
+    "@tableland/sdk": "^4.0.0-pre.6",
+    "@tableland/sqlparser": "1.0.0",
```
Comment (carsonfarmer) [inline package.json:72]: This will need to be updated, surprised it isn't causing issues already?

File: src/cli.ts [hunk 4]
```diff
@@ -30,6 +31,16 @@ const config = explorer.search();
 // If a dotenv file (or exported env vars) are provided, these override any config values
 dotenv.config();
 
+export type GlobalOptions = {
```
Comment (carsonfarmer) [inline src/cli.ts:34]: nice, good idea

File: src/cli.ts [hunk 5]
```diff
@@ -30,6 +31,16 @@ const config = explorer.search();
...
+  privateKey: string;
+  chain: helpers.ChainName;
+  providerUrl: string;
+  baseUrl: string;
+  verbose: boolean;
+  ensProviderUrl?: string;
```
Comment (carsonfarmer) [inline src/cli.ts:40]: could we just merge these into a single option, so if ensProviderUrl is provided, then ens is enabled?

File: src/commands/chains.ts [hunk 6]
```diff
@@ -1,14 +1,17 @@
 import type { Arguments, CommandBuilder } from "yargs";
+import { GlobalOptions } from "../cli.js";
 import { getChains } from "../utils.js";
+import type yargs from "yargs";
 
-export type Options = {
+export type Options = GlobalOptions & {
```
Comment (carsonfarmer) [inline src/commands/chains.ts:6]: You can probably just do a plain type alias here. Not sure why I left that "empty" before?

File: src/commands/controller.ts [hunk 7]
```diff
@@ -1,17 +1,12 @@
 import type yargs from "yargs";
 import type { Arguments, CommandBuilder } from "yargs";
-import { ChainName, Registry } from "@tableland/sdk";
+import { Registry } from "@tableland/sdk";
 import { getWalletWithProvider, getLink } from "../utils.js";
-export type Options = {
-  // Local
+import { GlobalOptions } from "../cli.js";
+
+export type Options = GlobalOptions & {
```
Comment (carsonfarmer) [inline src/commands/controller.ts:7]: I think this is an example where the interface stuff is nicer.

```typescript
export interface Options extends GlobalOptions {
  name: string;
  controller: string;
}
```

File: src/commands/info.ts [hunk 8]
```diff
@@ -18,29 +20,25 @@ export const builder: CommandBuilder<{}, Options> = (yargs) =>
...
 
-  if (!network) {
-    console.error("unsupported chain (see `chains` command for details)");
-    return;
-  }
+    if (argv.enableEnsExperiment && ens) {
+      name = await ens.resolveTable(name);
+    }
+
+    const parts = name.split("_");
```
Comment (carsonfarmer) [inline src/commands/info.ts:33]: Why are we doing this twice here? Also, isn't it still valid if there isn't a prefix?

File: src/commands/namespace.ts [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+export type Options = GlobalOptions & {
+  domain: string;
+  mappings: string[];
+  get: boolean;
+  set: boolean;
+};
+
+export const command = "namespace <domain> [mappings..]";
+export const desc = "Get info about a given table by name";
```
Comment (carsonfarmer) [inline src/commands/namespace.ts:14]: I think this might be wrong?

File: src/commands/namespace.ts [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+  if (!ens) return;
+
+  if (get) {
+    console.log(await ens.resolveTable(domain));
+  }
+
+  if (set) {
+    const records = mappings.map((entry: any) => {
+      const [key, value] = entry.split("=");
+      console.log(entry);
```
Comment (carsonfarmer) [inline src/commands/namespace.ts:44]: Leftover?

File: src/commands/read.ts [hunk 11]
```diff
@@ -1,19 +1,16 @@
...
-  // Local
+export type Options = GlobalOptions & {
   statement?: string;
   format: "pretty" | "table" | "objects";
   file?: string;
-
-  // Global
-  chain: ChainName;
-  baseUrl: string | undefined;
+  providerUrl: string;
```
Comment (carsonfarmer) [inline src/commands/read.ts:12]: aren't these already in the global options?

File: src/commands/read.ts [hunk 12]
```diff
@@ -63,8 +55,12 @@ export const handler = async (argv: Arguments<Options>): Promise<void> => {
       );
       return;
     }
+    const db = database;
```
Comment (carsonfarmer) [inline src/commands/read.ts:58]: Why the copy here?

File: src/lib/EnsResolver.ts [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+      });
+      return true;
+    } catch (e: any) {
+      console.log("Adding table to ENS failed");
+      console.error(e.message);
+    }
+    return false;
+  }
+
+  async resolve(statement: string): Promise<string> {
```
Comment (carsonfarmer) [inline src/lib/EnsResolver.ts:72]: So cool!

File: src/lib/commandSetup.ts [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+        ? new Validator({ baseUrl })
+        : Validator.forChain(chain);
+  }
+}
+
+export async function setupCommand(
+  argv: GlobalOptions,
+  options?: ConnectionsOptions
+) {
+  await init();
```
Comment (carsonfarmer) [inline src/lib/commandSetup.ts:110]: See an example in the sdk where we effectively just memoize this so you aren't calling it each time. I think actually, it is meowed in the parser lib perhaps now that I think of it. We should confirm.

File: src/lib/initialConnectors.ts [hunk 15]
```diff
@@ -0,0 +1,21 @@
+// TODO: this stuff was declared here but never being used, commenting out so lint can pass -JW
+// import { Provider } from "@ethersproject/abstract-provider";
```
Comment (carsonfarmer) [inline src/lib/initialConnectors.ts:2]: Why do we even need this?

File: test/schema.test.ts (hunk 1 above)
Comment (carsonfarmer) [reply test/schema.test.ts:30]: That's right, so actually I'm not sure why we need this test anytmore if that's the case. Ideally we'll still check that the cli is till pikcing up the right table name

File: test/schema.test.ts [hunk 16]
```diff
@@ -32,14 +36,18 @@ describe("commands/schema", function () {
...
     assert.calledWith(consoleError, "Not Found");
   });
 
   test("Schema passes with local-tableland", async function () {
-    const consoleLog = spy(console, "log");
-    await yargs(["schema", "healthbot_31337_1"]).command(mod).parse();
-    assert.calledWith(consoleLog, {
+    const consoleDir = spy(console, "dir");
+    await yargs(["schema", "healthbot_31337_1", "--chain", "local-tableland"])
```
Comment (carsonfarmer) [inline test/schema.test.ts:47]: Shouldn't chain still be optional here?

File: src/cli.ts (hunk 5 above)
Comment (awmuncy) [reply src/cli.ts:40]: I'm leaving it separate on purpose so that someone has to explicitly enable ENS experiment. I want it to be 100% clear it's an experiment, not just another option.

File: src/commands/controller.ts (hunk 7 above)
Comment (awmuncy) [reply src/commands/controller.ts:7]: Agreed, updated Options on them all to be interface instead of type.

File: package.json (hunk 3 above)
Comment (awmuncy) [reply package.json]: Wasn't causing issues, but upgraded to 1.0.3

File: src/lib/initialConnectors.ts (hunk 15 above)
Comment (awmuncy) [reply src/lib/initialConnectors.ts]: Vestigial file, I've removed it.

File: src/lib/commandSetup.ts (hunk 14 above)
Comment (awmuncy) [reply src/lib/commandSetup.ts:110]: This function is only called once on any given run.

File: test/info.test.ts (hunk 2 above)
Comment (awmuncy) [reply test/info.test.ts]: Oops. I've removed the skip. I've also removed the specifying of the chain, since the command infers that from the tablename.

File: test/schema.test.ts (hunk 16 above)
Comment (awmuncy) [reply test/schema.test.ts]: You are correct, it's optional, and no longer specified in the test in the newest version of this file.

File: test/schema.test.ts (hunk 1 above)
Comment (awmuncy) [reply test/schema.test.ts]: I've updated this, it is now testing the chain in the tableland.

**Review by asutula — APPROVED** [review]
I was reviewing this mostly just from the perspective of the SDK update, not the ENS support feature. Looks good, but left one comment about the version in package.json.

File: package.json [hunk 17]
```diff
@@ -1,11 +1,16 @@
 {
   "name": "@tableland/cli",
-  "version": "4.0.0-pre6.0",
+  "version": "4.0.0-pre7.0",
```
[the remark this answers is not in the document]
Comment (asutula) [reply package.json:3]: Oh, actually it looks like we have the new publish workflow in this repo. In that case, this should just be `0.0.0` and the version number to publish is passes as input to that workflow when manually triggered.

File: src/cli.ts (hunk 5 above)
Comment (carsonfarmer) [reply src/cli.ts:40]: ok I buy it

File: package.json (hunk 17 above)
[the remark this answers is not in the document]
Comment (carsonfarmer) [reply package.json:3]: Yeh let's update this to 0.0.0 @awmuncy. But actually, this is going to be a minor version bump on the 4.x release I think right? Just some fixes and a new API? Can your double check me on that @asutula and @awmuncy?

[the remark this answers is not in the document]
Comment (awmuncy) [reply package.json]: Yes, this should be just a minor version bump from the 4.x. Fixes + the ENS (non-breaking) changes.

File: src/commands/schema.ts [hunk 18]
```diff
@@ -18,34 +17,29 @@ export const builder: CommandBuilder<{}, Options> = (yargs) =>
...
+    const [tableId, chainId] = name.split("_").reverse();
 
-  const parts = name.split("_");
-  if (parts.length < 3) {
-    console.error(
-      "invalid table name (name format is `{prefix}_{chainId}_{tableId}`)"
-    );
-    return;
-  }
+    const parts = name.split("_");
```
Comment (carsonfarmer) [inline src/commands/schema.ts:24]: why twice?

File: test/create.test.ts [hunk 19]
```diff
@@ -217,11 +217,11 @@ describe("commands/create", function () {
     const privateKey = account.privateKey.slice(2);
     const consoleLog = spy(console, "log");
     const stdin = mockStd.stdin();
-    process.nextTick(() => {
+    setTimeout(() => {
```
Comment (carsonfarmer) [inline test/create.test.ts:220]: Why this change?

File: test/read.test.ts [hunk 20]
```diff
@@ -7,8 +7,9 @@ import * as mod from "../src/commands/read.js";
 import { wait } from "../src/utils.js";
 
 describe("commands/read", function () {
+  this.timeout(10000);
   before(async function () {
-    await wait(500);
+    await wait(5000);
```
Comment (carsonfarmer) [inline test/read.test.ts:12]: 5 seconds? that's a long delay? why is this needed?

**Review by carsonfarmer — APPROVED** [review]
Ok, nice improvements here. I think we can merge and release this! Going thru this PR, it does seem like it'd be nice to mock the ENS resolving stuff for tests. I looks like it would be pretty easy actually, with a custom resolver that satisfied our required interface?

File: src/utils.ts [hunk 21]
```diff
@@ -67,11 +69,18 @@ export function getWalletWithProvider({
...
   if (!provider) {
     // This will be significantly rate limited, but we only need to run it once
-    provider = getDefaultProvider(network);
+    provider = getDefaultProvider({ ...network, name: network.chainName });
   }
+
   if (!provider) {
     throw new Error("unable to create ETH API provider");
   }
+  (async () => {
```
Comment (carsonfarmer) [inline src/utils.ts:78]: This isn't going to be safe. We'll have to update this method to be async. That kinda sucks, but async is infectious, so we'll need to update all the calls to it as well. Bummer.

**Review by carsonfarmer — CHANGES_REQUESTED** [review]
Just updating my review based on the async issue in the utils.

File: test/create.test.ts (hunk 19 above)
Comment (awmuncy) [reply test/create.test.ts:220]: Before, the entire program up to this point in the testing process was synchronous, which meant "nextTick" always meant the CLI was ready to accept input. Because of some asynchronous stuff that happens in the setup command, it no longer means it's ready at the next tic. The has to do with event loop ordering, I don't recall exactly how.

File: src/utils.ts (hunk 21 above)
Comment (awmuncy) [reply src/utils.ts]: > Just updating my review based on the async issue in the utils.

Updated. The Connections (created during setup) class is now only usable once a ready check has completed, during which it checks chainId compatibility.

File: src/commands/namespace.ts [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+      if (valueRegex.exec(value) === null) {
+        throw new Error("Tablename should be a valid tableland table name");
+      }
+      return {
+        key,
+        value,
+      };
+    });
+
+    console.log("Domain", domain);
```
[the remark this answers is not in the document]
Comment (awmuncy) [reply src/commands/namespace.ts]: Excellent point. The namespace command now logs JSON, so it should now be pipable.

File: package.json [hunk 23]
```diff
@@ -61,8 +67,9 @@
     "typescript": "^4.8.4"
   },
   "dependencies": {
-    "@tableland/sdk": "^4.0.0-pre.4",
-    "@tableland/sqlparser": "^1.0.0",
+    "@ensdomains/ensjs": "3.0.0-alpha.39",
+    "@tableland/sdk": "^4.0.0-pre.6",
```
[the remark this answers is not in the document]
Comment (carsonfarmer) [reply package.json:71]: oh shoot, right!

File: src/commands/namespace.ts (hunk 9 above)
Comment (carsonfarmer) [reply src/commands/namespace.ts:14]: Did you see this @awmuncy ?

File: src/commands/namespace.ts (hunk 10 above)
Comment (carsonfarmer) [reply src/commands/namespace.ts:44]: Bump

File: src/utils.ts [hunk 24]
```diff
@@ -67,11 +69,17 @@ export function getWalletWithProvider({
...
     // This will be significantly rate limited, but we only need to run it once
-    provider = getDefaultProvider(network);
+    provider = getDefaultProvider({ ...network, name: network.chainName });
   }
+
   if (!provider) {
     throw new Error("unable to create ETH API provider");
   }
+
+  if ((await provider.getNetwork()).chainId !== network.chainId) {
```
Comment (carsonfarmer) [inline src/utils.ts:79]: nice, thanks!

File: test/create.test.ts (hunk 19 above)
Comment (carsonfarmer) [reply test/create.test.ts:220]: cool, I buy it

**Review by carsonfarmer — APPROVED** [review]
There are a few outstanding comments/questions left. While I don't think they need to hold up merging this, there are some that point out potential debugging artifacts and typos, etc that should probably be addressed.

File: src/commands/namespace.ts (hunk 22 above)
[the remark this answers is not in the document]
Comment (carsonfarmer) [reply src/commands/namespace.ts]: sick!

File: src/commands/namespace.ts (hunk 10 above)
Comment (awmuncy) [reply src/commands/namespace.ts]: Indeed, removed.

File: src/commands/namespace.ts (hunk 9 above)
Comment (awmuncy) [reply src/commands/namespace.ts]: Got it.
